repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -210,4 +193,18 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
 
     compact(render(resultInJson))
   }
+
+  private def executeWithJobGroup(executionCount: Int, code: String): String = {
+    Kind(interpreter.kind) match {
+      case Spark() | PySpark() | PySpark3() =>
+        val cmd = """sc.setJobGroup(%d, "Job group for statement %d")"""
+          .format(executionCount, executionCount)
+        executeCode(executionCount, cmd)
+      case SparkR() =>
+        // SparkR doesn't support statement cancelling, so doesn't need to set job group
```

Comment:
You can find `setJobGroup` here. http://spark.apache.org/docs/latest/api/R/index.html
And SparkContext.cancel can be called anywhere, not necessary in the same thread. So we can call SparkContext.cancel in scala side.